### ISSUE TITLE
Checked/Debug tests take much longer on Ubuntu/OSX than Windows

### ISSUE THREAD

**Issue by mmitche** [dotnet/coreclr#2985]
Seeing significant differences between the checked run times on Windows vs. Ubuntu/OSX.

Typical Windows: 47 mins (build/test)
Typical Ubuntu (when it doesn't time out): 1 hr 47 mins (just test)

Release times look to be in the same ballpark.

I've upped the timeouts for now, but this needs to be investigated.

/cc @LLITCHEV @gkhanna79

**Comment by gkhanna79** [comment]
@janvorli What are your thoughts?

**Comment by janvorli** [comment]
@gkhanna79  I've looked at the log from the Ubuntu run where I can see how much time it took to run each single test and put them all to Excel to see if there are any big outliers that would build up the time difference, but it is not the case. Seems like the time is distributed among most of the tests. Unfortunately, I can only get total times for each test suite for Windows, not for individual tests, so I cannot directly compare the results test by test.
It seems we need to do some profiling to get some sense.
@mmitche when you've said that "Release times look to be in the same ballpark.", did you mean that they are similar on both systems or that there is also a large difference like in debug?

**Comment by mmitche** [comment]
@janvorli Look to be about the same between Windows and OSX/Ubuntu (reasonable times).

I don't think this is a case of an outlier.  The logs don't show any big gaps where one test ran forever.  What I might do is run a few cases on windows vs. Linux and do some basic timing.

Matt

**Comment by AndyAyersMS** [comment]
@janvorli you can get per-test timings from the CI. See [this page](http://dotnet-ci.cloudapp.net/job/dotnet_coreclr/job/release_windows_nt/828/testReport/JIT/) for instance. You can even sort by duration, so we know for instance RngchkStress3 is the slowest test at about 2m37s.

**Comment by jasonwilliams200OK** [comment]
We can probably diagnose it by spotting the exact commit which caused this performance degradation using a git debugging command; [`git bisect`](https://git-scm.com/docs/git-bisect).

First, describe the criteria of failure script in project's root:

```sh
# in ./criterion.sh
T="$(date +%H)"  # capture the time in hours

./build.sh
./tests/runtest.sh

T="$(($(date +%H)-T))" # capture new time and take difference in hours

# if more than one hour spent, exit with positive status, else 0
if((T>1)); then exit 0; else exit 1; fi
```

Then:

```sh
git bisect start
git bisect good dd9c20c # some old sha1 of commit which was building fine
git bisect bad master
git bisect run ./criterion.sh
```

@AndyAyersMS, I think CI is failing Ubuntu build for all the recent PRs including https://github.com/dotnet/coreclr/pull/2945 for similar reasons.

**Comment by janvorli** [comment]
@jasonwilliams200OK I believe this is not a regression - we were not running the tests on debug builds before. But thank you for reminding me the git bisect anyways, I remember I've seen an article on it a long time ago and then forgotten about it.

**Comment by janvorli** [comment]
@mmitche I have taken a look at results of one of my PR and it seems that the debug windows times are very similar to the Linux checked ones:
http://dotnet-ci.cloudapp.net/job/dotnet_coreclr/job/debug_windows_nt_prtest/2164/testReport/

**Comment by janvorli** [comment]
And looking back, even in a PR 12 days ago, the windows debug run was still taking about the same time.

**Comment by janvorli** [comment]
@mmitche - there is something interesting. If you look at the testReport page on Windows debug build, you'll see times corresponding to about 2 hours total. But if you look at the console log, you'll see the total time to be close to 47 minutes you have mentioned. 
So now I think that the test results timing doesn't match just because we run multiple tests in parallel. If we ran then without the parallelism, they would take those 2hours.

**Comment by janvorli** [comment]
And watching the tests run on Linux, it seems we are not running them in parallel at all.
So that would explain the difference.

**Comment by mmitche** [comment]
@janvorli I checked this earlier and we are running in parallel.  There were like 10 processes running when I checked.

**Comment by janvorli** [comment]
Ok, so maybe that's just my local setup then.

**Comment by mmitche** [comment]
I think the issue is that we're not building coreclr with optimizations in checked.  Fixing this.

**Comment by janvorli** [comment]
@mmitche I don't think it is the case. I have tried to build the stuff locally with -O3 and the perf was virtually unchanged. I believe the difference is really in the DEBUG macro definition which pulls in a lot of code.
I have hand picked one of the tests that in release build takes 10s and in debug build 1min22s and run it under a profiler (oprofile) on my native Linux box to see if I can pinpoint some major contributor to the performance problem. Unfortunately, there was nothing like that. 97% was spent in the libcoreclr.so and here is the list of the first 10 contributors:
```
samples  %        image name               app name                 symbol name
172289    7.7234  libcoreclr.so            corerun                  MethodTable::GetFlag(MethodTable::WFLAGS_HIGH_ENUM) const
110952    4.9738  libcoreclr.so            corerun                  TypeHandle::IsTypeDesc() const
75655     3.3915  libcoreclr.so            corerun                  Object::RawGetMethodTable() const
60319     2.7040  libcoreclr.so            corerun                  WKS::my_get_size(Object*)
59908     2.6856  libcoreclr.so            corerun                  CHECK::EnterAssert()
56255     2.5218  libcoreclr.so            corerun                  MethodTable::GetBaseSize()
52793     2.3666  libcoreclr.so            corerun                  MethodTable::HasComponentSize() const
48459     2.1723  libcoreclr.so            corerun                  MethodTable::SanityCheck()
47792     2.1424  libcoreclr.so            corerun                  TypeHandle::AsTypeDesc() const
42500     1.9052  libcoreclr.so            corerun                  MethodTable::TestFlagWithMask(MethodTable::WFLAGS_LOW_ENUM, MethodTable::WFLAGS_LOW_ENUM) const
```

**Comment by mmitche** [comment]
@janvorli That's interesting.  So you forced it to O3 and got the same perf?  Any ideas why the Windows Checked doesn't show the same behavior?

If we have a really bad DEBUG define we can probably optimize it

**Comment by janvorli** [comment]
Yes, with -O3, the test run took virtually the same time. But I'll double check that to make sure I haven't made a mistake in the process.
Btw, the test was GC/Regressions/v2.0-rtm/494226/494226
I was going to do some more investigation and profiling, but I got distracted yesterday. I'll look more into it today.

**Comment by mmitche** [comment]
You would have had to manually add O3, the compiler overrides from clang-compiler-override and windows-compiler-override are all ignored except for the CLR defines.

**Comment by janvorli** [comment]
@mmitche Hmm, I have modified the option in the clang-compiler-override and it was not ignored:
Here is a command line of one of the clang invocations, you can see that the option is O3 and the build is debug.
```/usr/bin/clang++-3.5   -DAMD64 -DBIT64=1 -DBUILDENV_CHECKED=1 -DDBG_TARGET_64BIT=1 -DDBG_TARGET_AMD64 -DDBG_TARGET_AMD64=1 -DDBG_TARGET_AMD64_UNIX -DDBG_TARGET_WIN64=1 -DDEBUG -DDEBUGGING_SUPPORTED -DDISABLE_CONTRACTS -DFEATURE_APPDOMAIN_RESOURCE_MONITORING -DFEATURE_ARRAYSTUB_AS_IL -DFEATURE_ASYNC_IO -DFEATURE_BCL_FORMATTING -DFEATURE_COLLECTIBLE_TYPES -DFEATURE_CORECLR -DFEATURE_COREFX_GLOBALIZATION -DFEATURE_CORESYSTEM -DFEATURE_CORRUPTING_EXCEPTIONS -DFEATURE_DBGIPC_TRANSPORT_DI -DFEATURE_DBGIPC_TRANSPORT_VM -DFEATURE_EVENTSOURCE_XPLAT=1 -DFEATURE_EVENT_TRACE=1 -DFEATURE_EXCEPTIONDISPATCHINFO -DFEATURE_FRAMEWORK_INTERNAL -DFEATURE_HIJACK -DFEATURE_HOSTED_BINDER -DFEATURE_HOST_ASSEMBLY_RESOLVER -DFEATURE_ICASTABLE -DFEATURE_IMPLICIT_TLS -DFEATURE_ISYM_READER -DFEATURE_LEAVE_RUNTIME_HOLDER=1 -DFEATURE_LEGACYNETCF -DFEATURE_LEGACYNETCFFAS -DFEATURE_LEGACYNETCF_DBG_HOST_CONTROL -DFEATURE_LOADER_OPTIMIZATION -DFEATURE_MAIN_CLR_MODULE_USES_CORE_NAME -DFEATURE_MANAGED_ETW -DFEATURE_MANAGED_ETW_CHANNELS -DFEATURE_MERGE_CULTURE_SUPPORT_AND_ENGINE -DFEATURE_MERGE_JIT_AND_ENGINE -DFEATURE_MULTICOREJIT -DFEATURE_NORM_IDNA_ONLY -DFEATURE_PAL -DFEATURE_PAL_SXS -DFEATURE_PERFMAP -DFEATURE_PREJIT -DFEATURE_RANDOMIZED_STRING_HASHING -DFEATURE_READYTORUN -DFEATURE_STANDALONE_SN -DFEATURE_STRONGNAME_DELAY_SIGNING_ALLOWED -DFEATURE_STRONGNAME_MIGRATION -DFEATURE_STUBS_AS_IL -DFEATURE_SVR_GC -DFEATURE_SYMDIFF -DFEATURE_SYNTHETIC_CULTURES -DFEATURE_UNIX_AMD64_STRUCT_PASSING -DFEATURE_UNIX_AMD64_STRUCT_PASSING_ITF -DFEATURE_USE_ASM_GC_WRITE_BARRIERS -DFEATURE_VERSIONING -DFEATURE_WIN32_REGISTRY -DFEATURE_WINDOWSPHONE -DFEATURE_WINMD_RESILIENT -DLINUX64 -DPLATFORM_UNIX=1 -DPROFILING_SUPPORTED -DUNICODE -DUNIX_AMD64_ABI -DURTBLDENV_FRIENDLY=Checked -DWRITE_BARRIER_CHECK=1 -D_AMD64_ -D_BLD_CLR -D_DBG -D_DEBUG -D_SECURE_SCL=0 -D_TARGET_AMD64_=1 -D_UNICODE -D_WIN64 -D__LINUX__=1 -Wall -Wno-null-conversion -std=c++11 -nostdinc++ -g -O3 -I/home/janvorli/git/coreclr/bin/obj/Linux.x64.Debug/src/vm/wks -I/home/janvorli/git/coreclr/src/vm/wks -I/home/janvorli/git/coreclr/src/vm -I/home/janvorli/git/coreclr/src/pal/prebuilt/inc -I/home/janvorli/git/coreclr/src/pal/inc -I/home/janvorli/git/coreclr/src/pal/inc/rt -I/home/janvorli/git/coreclr/src/pal/src/safecrt -I/home/janvorli/git/coreclr/src/inc -I/home/janvorli/git/coreclr/src/strongname/inc -I/home/janvorli/git/coreclr/src/inc/winrt -I/home/janvorli/git/coreclr/src/debug/inc -I/home/janvorli/git/coreclr/src/debug/inc/amd64 -I/home/janvorli/git/coreclr/src/debug/inc/dump -I/home/janvorli/git/coreclr/src/md/inc -I/home/janvorli/git/coreclr/src/classlibnative/bcltype -I/home/janvorli/git/coreclr/src/classlibnative/cryptography -I/home/janvorli/git/coreclr/src/classlibnative/inc -I/home/janvorli/git/coreclr/bin/obj/Linux.x64.Debug/src/inc -I/home/janvorli/git/coreclr/src/pal/inc/rt/cpp -I/home/janvorli/git/coreclr/src/nativeresources -I/home/janvorli/git/coreclr/src/gc -I/home/janvorli/git/coreclr/src/vm/amd64    -fno-omit-frame-pointer -fms-extensions -fstack-protector-strong -ferror-limit=4096 -Werror -Wno-unused-private-field -Wno-unused-variable -Wno-microsoft -Wno-tautological-compare -Wno-constant-logical-operand -Wno-unknown-warning-option -Wno-invalid-offsetof -Wno-incompatible-ms-struct -nostdinc -fPIC -o CMakeFiles/cee_wks.dir/__/__/gc/gccommon.cpp.o -c /home/janvorli/git/coreclr/src/gc/gccommon.cpp
```

**Comment by mmitche** [comment]
What cmake version is this?  That might explain some things.

**Comment by janvorli** [comment]
@mmitche 
I am using the CMake 2.8.12.2 which is the version provided in the Ubuntu 14.04.

I am sorry, it looks I made some mistake yesterday - I've tried it again today and the optimization option actually does have an effect. With O3, run of the debug build of the test took 15s (compared to 1min22s with O0). As I've said, the release build was taking 10s.
Here are the results for the test for O0, O1, O2 and O3 in debug builds:

 Optimization | Time [s]
-----------------|-------
O0 | 82
O1 | 37
O2 | 15
O3 | 15

**Comment by mmitche** [comment]
I'm on the exact same version.  I added -notaflag to DEBUG_INIT and didn't see it at all.  Can you push your changes to a branch so I an verify?

**Comment by janvorli** [comment]
I will, but let me try to experiment a little bit more. I was modifying the setting for the Debug build, so I wonder if we have a bug somewhere in the cmake files that causes that these options are ignored for the Checked build.

**Comment by janvorli** [comment]
It seems to work ok for the checked build as well.

**Comment by LLITCHEV** [comment]
@janvorli That is good analysis and numbers above. It looks like the O1 v. release build has a bit more than doubled the execution time. I assume this doesn't include the DEBUG flag passed to the build. I think, it 'd be useful to know what the O1 number will be if we also add the DEBUG flag. Do you have such number?

**Comment by janvorli** [comment]
@LLITCHEV The numbers are with DEBUG flag defined. I was just modifying the option for regular debug build.

**Comment by LLITCHEV** [comment]
@janvorli Would it be interesting to see how much delay is introduced by the O1 v. O2 optimization options on clang and the extra code introduced by the DEBUG ifdef? If it is not a big thing to get these, I think it might give interesting clues...

**Comment by janvorli** [comment]
Fixed by #3033
